readthedocs/search/api/v3/views.py

```diff
@@ -0,0 +1,134 @@
+from readthedocs.search.api.v3.backend import Backend
+from readthedocs.search.api.v3.serializers import PageSearchSerializer
+from readthedocs.search.api.v3.utils import should_use_advanced_query
+
+log = structlog.get_logger(__name__)
+
+
+class SearchAPI(GenericAPIView):
+
+    """
+    Server side search API V3.
+
+    Required query parameters:
+
+    - **q**: [Search term](https://docs.readthedocs.io/page/server-side-search/syntax.html).
+
+    Check our [docs](https://docs.readthedocs.io/page/server-side-search/api.html) for more information.
+    """  # noqa
+
+    http_method_names = ["get"]
+    pagination_class = SearchPagination
+    serializer_class = PageSearchSerializer
+    search_backend_class = Backend
+
+    def get_view_name(self):
+        return "Search API V3"
+
+    def _validate_query_params(self):
+        query = self.request.GET.get("q")
+        errors = {}
+        if not query:
+            errors["q"] = [_("This query parameter is required")]
+        if errors:
+            raise ValidationError(errors)
+
+    @cached_property
+    def _backend(self):
+        backend = self.search_backend_class(
+            request=self.request,
+            query=self.request.GET["q"],
+        )
+        return backend
+
+    def _get_search_query(self):
+        return self._backend.parser.query
+
+    def _get_projects_to_search(self):
+        return self._backend.projects
+
+    def get_queryset(self):
+        """
+        Returns an Elasticsearch DSL search object or an iterator.
+
+        .. note::
+
+           Calling ``list(search)`` over an DSL search object is the same as
+           calling ``search.execute().hits``. This is why an DSL search object
+           is compatible with DRF's paginator.
+        """
+        use_advanced_query = should_use_advanced_query(self._get_projects_to_search())
+        search = self._backend.search(
+            use_advanced_query=use_advanced_query,
+            aggregate_results=False,
+        )
+        if not search:
+            return []
+
+        return search
+
+    def get(self, request, *args, **kwargs):
+        self._validate_query_params()
+        result = self.list()
+        self._record_query(result)
+        return result
+
+    def _record_query(self, response):
+        total_results = response.data.get("count", 0)
+        time = timezone.now()
+        query = self._get_search_query().lower().strip()
+        # NOTE: I think this may be confusing,
+        # since the number of results is the total
+        # of searching on all projects, this specific project
+        # could have had 0 results.
+        for project, version in self._get_projects_to_search():
+            tasks.record_search_query.delay(
+                project.slug,
+                version.slug,
+                query,
+                total_results,
+                time.isoformat(),
+            )
+
+    def list(self):
```

Comment:
Would be useful to understand why we're overriding this method. A comment/docstring would be good.

Reply:
This isn't being overridden, it just a method of this class p: The confusion may be bc the `ListAPIView` has a list method, but this is just a `GenericView`, but we are using it to list search results, so I used a similar structure.

Reply:
Good to note that then, also :)